File: LeetCode/DataStruct/Golang/Leetcode895-maximum-frequency-stack.go
package sbgo

type FreqStack struct {
	mx   int
	freq map[int]int
	rec  map[int][]int
}

func max(i, j int) int {
	if i < j {
		return j
	}
	return i
}

func Constructor() FreqStack {
	return FreqStack{0, map[int]int{}, map[int][]int{}}
}

func (this *FreqStack) Push(val int) {
	this.freq[val]++
	this.rec[this.freq[val]] = append(this.rec[this.freq[val]], val)
	this.mx = max(this.mx, this.freq[val])
}

func (this *FreqStack) Pop() int {
	n := len(this.rec[this.mx])
	res := this.rec[this.mx][n-1]
	this.freq[res]--
	this.rec[this.mx] = this.rec[this.mx][:n-1]
	if n-1 == 0 {
		this.mx -= 1
	}
	return res
}

File: LeetCode/DataStruct/Python/Leetcode895-maximum-frequency-stack.py
class FreqStack:

    def __init__(self):
        self.rec = defaultdict(list)
        self.freq = defaultdict(int)
        self.mx = 0

    def push(self, val: int) -> None:
        self.freq[val] += 1
        self.mx = max(self.mx, self.freq[val])
        self.rec[self.freq[val]].append(val)

    def pop(self) -> int:
        res = self.rec[self.mx].pop()
        self.freq[res] -= 1
        if not self.rec[self.mx]: 
            self.mx -= 1
        return res


# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()